### PR TITLE
Requiring authentication for user data queries and mutations

### PR DESCRIPTION
**Pull request by MChartier**
Requiring authentication for most queries/mutations and limiting access
to only the authenticated user's data. User ID is taken  #from the
caller's token instead of being provided as input.

The exceptions are the 'user' query and 'createUser' mutation which
remain for development and testing purposes; these must be restricted input
Production environments before launch.

Adding environment variable configuration to allow developers to
manually configure a test user identity in non-production environments
to simplify testing. Rather than creating a test user and logging in to
get an access token, token validation is bypassed and the user ID may be
read from an environment variable, though it must begin with 'test|'.
(Note: requests will still fail if an INVALID token is provided, even in
a non-production environment; but requests without a token will succeed
if test data is configured for use instead.)

#28

### CODE REVIEW

File: auth/authenticate.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	sub, ok := claims["sub"].(string)
+	if !ok {
+		return nil, errors.New("unable to read 'sub' value from user token")
+	}
+
+	return &sub, nil
+}
+
+func validationKeyGetter(token *jwt.Token) (interface{}, error) {
+	// Verify 'aud' claim
```
Comment (SlyCaptainFlint) [inline auth/authenticate.go:122]: what is aud and iss?

File: graph/schema.graphqls [hunk 2]
```diff
@@ -59,8 +57,9 @@ input NewUser {
 }
 
 type Mutation {
+  createMe: User!
```
Comment (SlyCaptainFlint) [inline graph/schema.graphqls:60]: how is this mutation meant to work without any variables?

File: graph/schema.resolvers.go [hunk 3]
```diff
@@ -10,21 +10,39 @@ import (
...
+
+	// Try to get user with given token sub as ID
+	// This will fail if the user already exists
+	return r.CreateUser(ctx, model.NewUser{
+		ID: *tokenSub,
+	})
+}
+
 func (r *mutationResolver) CreateUser(ctx context.Context, input model.NewUser) (*model.User, error) {
+	// TODO: Require elevated permissions to create user other than the one who is authenticated
```
Comment (SlyCaptainFlint) [inline graph/schema.resolvers.go:35]: still to do?

File: graph/schema.resolvers.go [hunk 4]
```diff
@@ -10,21 +10,39 @@ import (
...
+func (r *mutationResolver) CreateMe(ctx context.Context) (*model.User, error) {
+	// Get the user ID from context
+	// If not found, request is not authenticated and should fail
+	tokenSub, err := auth.GetUserIDFromContext(ctx)
+	if tokenSub == nil || err != nil {
+		return nil, errors.New("request failed: caller not authenticated")
+	}
+
+	// Try to get user with given token sub as ID
+	// This will fail if the user already exists
```
Comment (SlyCaptainFlint) [inline graph/schema.resolvers.go:28]: if the user does already exist, what response is the client going to get?

File: auth/authenticate.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	jwtmiddleware "github.com/auth0/go-jwt-middleware"
+	"github.com/dgrijalva/jwt-go"
+)
+
+const auth0ApiIdentifier = "matterhorn-api"
+const auth0Domain = "https://matterhorn-prototype.auth0.com/"
+
+// getPemCerts gets the remote JWKS for the Auth0 account and returns the certificate with the public key in
+// PEM format
+func getPemCert(token *jwt.Token) (string, error) {
```
Comment (SlyCaptainFlint) [inline auth/authenticate.go:22]: what is Pem?

File: auth/authenticate.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	claims, ok := userToken.Claims.(jwt.MapClaims)
+	if claims == nil || !ok {
+		return nil, errors.New("unable to parse claims on user token")
+	}
+
+	sub, ok := claims["sub"].(string)
+	if !ok {
+		return nil, errors.New("unable to read 'sub' value from user token")
```
Comment (SlyCaptainFlint) [inline auth/authenticate.go:115]: what is this "sub" value?

File: auth/authenticate.go (hunk 1 above)
Comment (SlyCaptainFlint) [inline auth/authenticate.go:122]: what is aud and iss?

Comment (MChartier) [reply auth/authenticate.go:122]: "Audience" and "Issuer". Will add some clarifying comments...

File: environment/environment.go [hunk 7]
```diff
@@ -13,7 +13,7 @@ func LoadEnv(relPath string) {
 	env := os.Getenv("MATTERHORN_ENV")
 	if "" == env {
 		// Default to "local"
-		env = "dev"
+		env = "local"
```
Comment (SlyCaptainFlint) [inline environment/environment.go]: I feel like defaulting to non-prod if the env is not defined is a bit dangerous considering that removes the need to authenticate for a bunch of different request types. Consider throwing an exception if env is not defined.

File: graph/schema.resolvers.go (hunk 3 above)
Comment (MChartier) [reply graph/schema.resolvers.go:35]: Yes. This is still exposed as a mutation endpoint and we don't enforce any kind of authorization to access it.

File: auth/authenticate.go (hunk 1 above)
Comment (MChartier) [reply auth/authenticate.go:122]: "Audience" and "Issuer". Will add some clarifying comments...

**Review by SlyCaptainFlint — APPROVED** [review]

File: environment/environment.go (hunk 7 above)
Comment (MChartier) [reply environment/environment.go]: Good catch, this doesn't account for the fact that there are values being set as environment variables. Will change this.

File: graph/schema.graphqls (hunk 2 above)
Comment (MChartier) [reply graph/schema.graphqls:60]: Creates a user with the ID encoded in the caller's access token.

File: graph/schema.resolvers.go (hunk 4 above)
Comment (MChartier) [reply graph/schema.resolvers.go:28]: Error response. (Line 45)

File: auth/authenticate.go (hunk 6 above)
Comment (MChartier) [reply auth/authenticate.go:115]: "Subject". It's the unique user ID assigned by Auth0. Will add clarifying comments...

File: auth/authenticate.go (hunk 5 above)
Comment (MChartier) [reply auth/authenticate.go:22]: Something Encryption Something. This is copy-pasted from Auth0, will find out and add more useful comments...

**Pull request merged**